lib/srv/db/postgres/users.go

```diff
@@ -205,33 +198,26 @@ func (e *Engine) applyPermissions(ctx context.Context, sessionCtx *common.Sessio
 		return trace.BadParameter("fine-grained database permissions and database roles are mutually exclusive, yet both were provided.")
 	}
 
-	rules, err := e.AuthClient.GetDatabaseObjectImportRules(ctx)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	conn, err := e.connectAsAdmin(ctx, sessionCtx, false)
+	fetcher, err := objects.GetObjectFetcher(ctx, sessionCtx.Database, objects.ObjectFetcherConfig{
```

Comment:
I really like reusing the same fetcher for user sessions. Is it possible to reuse the same admin connection for calling the permission procedures though?

Reply:
Reuse sounds tricky to do properly. I'll look into it, I do have an idea that might work. Even if not, perhaps this is acceptable for the ease of maintenance this otherwise brings.

Reply:
could we cache it in the `fetcher`? something like

```
type cachedAdminPGXConnGetter interface {
  getCachedPGXAdminConn() (*pgx.Conn, error)
}

if cached, ok := fetcher.(cachedAdminPGXConnGetterr); ok {
  conn, err = cached.getCachedAdminConn()
} else {
  conn, err = e.connectAsAdminSessionDatabase(ctx, sessionCtx)
} 
```
Or maybe move the apply part to the fetcher interface as well. I will leave it up to you if you want to keep it like this or we can tackle this later when doing performance improvement.

Reply:
Yep, I've tried something along these lines. Trouble is with ensuring the connection params are the same, which often isn't the case. I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

Reply:
>  I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

I just gave it another thought that it won't work well as we have to deal with closing the connection.

Reply:
Yes, that is another complication I've run into. I've considered using a connection pool instead, but those are not designed for our use case at all.